[RNMobile] Upgrade React Native `0.71.11` - Android changes

fluiddot opened the pull request:
**Related PRs:**
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5872
* https://github.com/wordpress-mobile/WordPress-Android/pull/18613

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Applies the needed changes to Android native files to support the upgrade of React Native to version ~~`0.71.8`~~`0.71.11`.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This PR is needed as part of the effort of upgrading React Native to version ~~`0.71.8`~~`0.71.11`. 

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Most of the changes have been applied following [the React Native upgrade helper](https://react-native-community.github.io/upgrade-helper/?from=0.69.4&to=0.71.11). Here is a list of the different commits applied and some notes:

* [Use React Native and Hermes modules from Maven](https://github.com/WordPress/gutenberg/commit/d108e4d70456c87ba0cdb953951e1596ad81072f)
* [Add `mavenLocal` repository to allow testing local binaries](https://github.com/WordPress/gutenberg/commit/a9f99ea98cc871c2656a07d52b7323dfad0f8fdc)
* [Bump Reanimated and Gesture handler libraries](https://github.com/WordPress/gutenberg/commit/05b94d1ff57027d53e4fed98a42cf80ca65f42aa)
* Tweaks to build configuration of `react-native-editor`
  * [Remove gradle download task plugin](https://github.com/WordPress/gutenberg/commit/2b366c186e17bb7b8c9489594da68cd428975431)
  * [Bump ndk version](https://github.com/WordPress/gutenberg/commit/ddb7145586752013d4481cb809e881bc5f15b036)
  * [Remove no longer need logic related to the new architecture](https://github.com/WordPress/gutenberg/commit/86b22be20f3d20d2655de3a1bc26de91e14b55ea)
  * [Remove deprecated Gradle property](https://github.com/WordPress/gutenberg/commit/72ac85315474f6f320b6164020936fdbc87ebff1)
* Updates to the Gutenberg demo app
  * [Remove no longer needed files in new version](https://github.com/WordPress/gutenberg/commit/13291dea896beedb577501685980402dd90a4786)
  * [Update Flipper initialization](https://github.com/WordPress/gutenberg/commit/20f783bbe2e81f70b9e1608a934215ef7ab0f6f4)
  * [Update main application](https://github.com/WordPress/gutenberg/commit/6de2bc6822444e80757a3514bac731759bbc2702)
  * [Apply plugin React Native Gradle plugin](https://github.com/WordPress/gutenberg/commit/d4159c2a51bf7a563807c453c1d52c44cbcac912)
  * [Resolve react-native-gradle-plugin incompatability](https://github.com/WordPress/gutenberg/commit/437450efd009e0d953369eea47f5ca972abef1f7)
  * [Remove exclude group from Flipper](https://github.com/WordPress/gutenberg/commit/873a345f907ada8e4b1d43206c62480ed8f0e450)
  * [Update comments in `build.gradle` to align with new RN version](https://github.com/WordPress/gutenberg/commit/1bd3d3e41707f3d195f97323eb3dbecd7fc2a7cb)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
Follow testing instructions from https://github.com/WordPress/gutenberg/pull/51303.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->
N/A

fluiddot committed:
Upgrade Gradle to version 7.5.1

fluiddot committed:
Upgrade Gradle plugin

fluiddot committed:
Remove no longer needed files in new version

fluiddot committed:
Update Flipper initialization

fluiddot committed:
Update demo project main application

fluiddot committed:
Remove gradle download task plugin

fluiddot committed:
Bump ndk version

fluiddot committed:
Remove no longer need logic related to `newArchEnabled`

fluiddot committed:
Apply plugin React Native Gradle plugin

fluiddot committed:
Use React Native and Hermes modules from Maven

We no longer need to publish these binaries because React Native team is publishing React Native binaries to Maven.

fluiddot committed:
Remove exclude group from Flipper

fluiddot committed:
Update comments in `build.gradle` to align with new RN version

fluiddot committed:
Remove deprecated Gradle property

fluiddot committed:
Add `mavenLocal` repository to allow testing local binaries

fluiddot committed:
Bump Reanimated and Gesture handler libraries

dcalhoun commented on packages/react-native-aztec/android/gradlew:
```diff
@@ -205,6 +205,12 @@ set -- \
         org.gradle.wrapper.GradleWrapperMain \
         "$@"
 
+# Stop when "xargs" is not available.
+if ! command -v xargs >/dev/null 2>&1
+then
+    die "xargs is not available"
+fi
+
```
I was unable to locate this and other Gradle changes in the the React Native Upgrade Helper. Were these required by something else?

fluiddot replied:
This file was automatically created when updating the Gradle wrapper (https://github.com/WordPress/gutenberg/pull/51289/commits/d108e4d70456c87ba0cdb953951e1596ad81072f). The wrapper is present in the three RN packages `react-native-aztec`, `react-native-bridge`, and `react-native-editor`. I think we don't see it in the upgrade helper because these files are ignored ([reference](https://github.com/facebook/react-native/blob/d097408c1ca15d3aa9dd69dc12f1043f54890213/.gitignore#L33-L39)).

dcalhoun commented:
Side note: force pushing apparently breaks the URLs to a Pull Request's commits (found in a Pull Request commits tab), which means the links in this PR description's "How?" section are now broken. URLs to the unnamed commits themselves (traditional commit URLs) remain working, as GitHub appears to never garbage collect unnamed commits, unlike like Git itself. So, in theory, the unnamed commit URLs remain operational forever. 

So, it may be best to use that URL or merely paste the commit SHA itself, which GitHub turns into a URL, e.g. b82d54f72a5ac64cd6614e3d0a1540d70b321fdf.

| Status | Commit | PR Commit URL | Traditional Commit URL |
| - | - | - | - |
| [Unnamed](https://github.com/WordPress/gutenberg/commit/14932d7c93611f3b04e32f6dd56ebf78c612b94a) | `14932d7c93611f3b04e32f6dd56ebf78c612b94a` | `https://github.com/WordPress/gutenberg/pull/51289/commits/14932d7c93611f3b04e32f6dd56ebf78c612b94a` | `https://github.com/WordPress/gutenberg/commit/14932d7c93611f3b04e32f6dd56ebf78c612b94a` |
| [Named](https://github.com/WordPress/gutenberg/commit/b82d54f72a5ac64cd6614e3d0a1540d70b321fdf) | `b82d54f72a5ac64cd6614e3d0a1540d70b321fdf ` | `https://github.com/WordPress/gutenberg/pull/51289/commits/b82d54f72a5ac64cd6614e3d0a1540d70b321fdf` | `https://github.com/WordPress/gutenberg/commit/b82d54f72a5ac64cd6614e3d0a1540d70b321fdf` |

fluiddot commented:
> Side note: force pushing apparently breaks the URLs to a Pull Request's commits (found in a Pull Request commits tab), which means the links in this PR description's "How?" section are now broken. URLs to the unnamed commits themselves (traditional commit URLs) remain working, as GitHub appears to never garbage collect unnamed commits, unlike like Git itself. So, in theory, the unnamed commit URLs remain operational forever.

Ah, my bad. I'll update the broken link 💨 . Ah, good idea about using unnamed commits, although they will show the commit orphaned at least it won't be broken. In any case, I'll update the links with the latest commits. Thanks 🙇 !

fluiddot committed:
Merge branch 'rnmobile/upgrade/react-native-0.71.8' into rnmobile/upgrade/react-native-0.71.8-android

dcalhoun committed:
build: Resolve react-native-gradle-plugin incompatability

Due to host app requirements, we must use AGP 7.2.1. The included patch
disables logic requiring AGP 7.3. The logic appears to not be required
for our use cases in the Demo editor or host apps.

We should remove this patch once we upgrade past AGP 7.3.

dcalhoun committed:
Revert "Upgrade Gradle plugin"

This reverts commit 82764a265d6b8ea57ea6b6c420b3e244d347ec03.

dcalhoun committed:
Merge branch 'rnmobile/upgrade/react-native-0.71.8' of github.com:WordPress/gutenberg into rnmobile/upgrade/react-native-0.71.8-android

fluiddot committed:
Reduce priority of `mavenLocal` repository in Android build configurations

Maven local is used to provide dependencies located locally, which is mainly used for testing and debugging. Hence, published dependencies should be prioritized over local ones.

fluiddot committed:
Merge branch 'rnmobile/upgrade/react-native-0.71.8' into rnmobile/upgrade/react-native-0.71.8-android

fluiddot committed:
Disable `react-native-screens` in navigators

`react-native-screens` is meant to be used at root level to save memory when having inactive screens. This is not the case of the editor, as the stack navigators are used within the Bottom sheet component.
As a side note, enabling `react-native-screens` here leads to the editor crashing.

fluiddot committed:
Fix render order of animated view to highlight selected segment

Rendering the animated view before the segments will ensure that is rendered behind them.

fluiddot commented on packages/components/src/mobile/segmented-control/index.native.js:
```diff
@@ -149,6 +149,17 @@ const SegmentedControls = ( {
 		<View style={ styles.row }>
 			<View style={ styles.flex }>{ addonLeft }</View>
 			<View style={ [ containerStyle, isIOS && styles.containerIOS ] }>
+				<Animated.View
+					style={ [
+						{
+							width,
+							left: positionAnimationValue,
+							height,
+						},
+						selectedStyle,
+						outlineStyle,
+					] }
+				/>
```
I had to change the rendering order after upgrading `@react-navigation/stack` to version `6.3.5` (https://github.com/WordPress/gutenberg/pull/51303/commits/b5c02fd4936dd2dfd63cc73fd0ed807871ed05f0). Otherwise, the animated component was being rendered over the segments despite the `z-index` value.

| Before the fix | After the fix |
|--------|--------|
| <img src=https://github.com/WordPress/gutenberg/assets/14905380/59998f5b-de88-4c12-b3a7-cb455c6646d3 width=300> | <img src=https://github.com/WordPress/gutenberg/assets/14905380/c35c3253-6f87-4dcf-8f9b-24158abcd803 width=300> |

fluiddot committed:
Merge branch 'rnmobile/upgrade/react-native-0.71.8' into rnmobile/upgrade/react-native-0.71.8-android

fluiddot committed:
Bump Linear gradient Android library

This library is now published via the `react-native-libraries-publisher` repository.

fluiddot committed:
Merge branch 'rnmobile/upgrade/react-native-0.71.8' into rnmobile/upgrade/react-native-0.71.8-android

fluiddot committed:
Merge branch 'rnmobile/upgrade/react-native-0.71.8' into rnmobile/upgrade/react-native-0.71.8-android

SiobhyB reviewed (approved):
LGTM! Verified I'm able to build the demo app with https://github.com/wordpress-mobile/gutenberg-mobile/pull/5872 also checked out, and the changes match those provided by the RN upgrade helper. 👏 👏 🧙
